Python/ceval.c

```diff
@@ -207,8 +207,11 @@ _PyEval_FiniThreads(void)
 static inline void
 exit_thread_if_finalizing(PyThreadState *tstate)
 {
+    /* Get interpreter pointer */
+    PyInterpreterState *interp = tstate->interp;
+
     /* _Py_Finalizing is protected by the GIL */
-    if (_Py_IsFinalizing() && !_Py_CURRENTLY_FINALIZING(tstate)) {
```

Comment:
With this change, _Py_CURRENTLY_FINALIZING() becomes even more weird that it was: you rely on a global and unique PyRuntimeState to check which thread is currently being finalized. I don't think that it's correct. It should be possible to finalize 2 interpreters in parallel which both have their own thread.

For me, it would make more sense to move "PyThreadState *finalizing;" from _PyRuntimeState to PyInterpreterState and remove PyInterpreterState.finalizing.

What's your call Eric? @ericsnowcurrently

Reply:
The thread state in `finalizing` actually means "This is the thread that is finalizing the runtime", so other threads can easily tell that another thread is responsible for cleaning things up.

Reply:
However @pablogsal pointed out a more serious problem here, which is that simply moving the existing check will result in:

* potentially causing threads belonging to other subinterpreters to exit
* potentially causing threads belonging to the embedding application to exit

The latter problem already exists, but moving the cleanup to EndInterpreter makes it easier to hit.